CMakeLists.txt

```diff
@@ -4,10 +4,63 @@
 
 cmake_minimum_required(VERSION 2.8)
 
+option(USE_CPP03 "use cpp 03 standard" OFF)
+
+# cmake build type
 if(DEFINED CMAKE_BUILD_TYPE)
-   set(CMAKE_BUILD_TYPE ${CMAKE_BUILD_TYPE} CACHE STRING "Debug Release RelWithDebInfo MinSizeRel")
+	set(CMAKE_BUILD_TYPE_DEFAULT ${CMAKE_BUILD_TYPE})
+else()
+	set(CMAKE_BUILD_TYPE_DEFAULT RelWithDebInfo)
+endif()
+set(CMAKE_BUILD_TYPE ${CMAKE_BUILD_TYPE_DEFAULT} CACHE STRING "build type")
+set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS Debug Release RelWithDebInfo MinSizeRel)
+
+# try to determine default platform from cmake system name
+if(DEFINED CMAKE_SYSTEM_NAME)
+	if(${CMAKE_SYSTEM_NAME} STREQUAL "Linux")
+		set(OS_DEFAULT "posix")
+	endif()
 else()
-   set(CMAKE_BUILD_TYPE RelWithDebInfo CACHE STRING "Debug Release RelWithDebInfo MinSizeRel")
+	set(OS_DEFAULT "generic")
+endif()
+
+# define config option
+set(BOARD generic CACHE STRING "board configuration")
+set_property(CACHE BOARD PROPERTY STRINGS
+	stm32 generic)
+
+# define os option
+set(OS generic CACHE STRING "operating system")
+set_property(CACHE OS PROPERTY STRINGS
+	posix chibios nuttx baremetal generic)
+string(TOUPPER ${OS} OS_UPPER)
+
+# stm32 ifaces option
+set(UAVCAN_STM32_NUM_IFACES "1" CACHE STRING "number of interfaces")
+set_property(CACHE UAVCAN_STM32_NUM_IFACES PROPERTY STRINGS 1 2)
+
+# nuttx export
+set(NUTTX_EXPORT_DIR "" CACHE PATH "nuttx export directory")
+
+# board/os based flags
+if (${BOARD} STREQUAL "generic")
+elseif (${BOARD} STREQUAL "stm32")
+	if (${OS} STREQUAL "nuttx")
+		if("${NUTTX_EXPORT_DIR}" STREQUAL "")
+			message(FATAL_ERROR "please define the nuttx export directory")
+		endif()
+	endif()
+	include_directories(${NUTTX_EXPORT_DIR}/include)
+	link_directories(${NUTTX_EXPORT_DIR}/lib)
+	link_libraries(
+		nosys
+		nuttx	
+		apps
```

Comment:
Isn't this set of libraries specific to PX4 configuration, and may not be valid for a generic NuttX build?